src/NUnitFramework/framework/Internal/Builders/DefaultTestCaseBuilder.cs

```diff
@@ -132,11 +132,30 @@ public Test BuildFrom(FixtureMethod method, Test suite)
                     tests.Add(test); 
             }
 
-            return hasBuildersSpecified && method.Method.GetParameters().Length > 0 || tests.Count > 0
+            return hasBuildersSpecified && method.Method.GetParameters().Length > 0 || 
+                   tests.Count > 0 ||
+                   MethodHasDataAttributeParameters(method)
```

Comment:
DataAttribute is a class that was somewhat important in earlier versions of NUnit. Currently, it's a holdover and is not used in the code at all. We left it in the code to avoid needlessly breaking anything. 

This change, if it were to be implemented would lock in `DataAttribute` as required for all data classes that apply to parameters. User classes would be required to inherit from it and any existing user attributes that __do not__ inherit from it would break.

I don't particularly like this approach of scanning for attributes (actually re-scanning since we do it elsewhere) but if it's the only solution to the problem, you should look for the `IParameterDataSource` interface rather than `DataAttribute`.

Reply:
I think what we need to ask here is "Why is `hasBuilderSpecified` insufficient?" Of course, the answer is that we give the method a hidden `CombinatorialAttribute` if it doesn't already have one of the combining attributes on it. We do so __whether or not__ any parameter data sources are provided. If we are going to check for parameter data sources, perhaps we should do it up front and only add the unspecified `CombinatorialAttribute` if it's __really__ needed. Then `hasBuilderSpecified` could be set correctly.

Reply:
I like this approach. If I understand correctly, we can implement this by moving `var hasBuildersSpecified =` to after the `if (needCombinatorial)` statement.

@CharliePoole Beyond moving that statement: we shouldn't be checking for parameter attributes as builders, should we? Discovering parameter attributes as builders seems like a bigger change which would be outside the scope of this PR.